discovery/src/main/java/io/airlift/discovery/client/DiscoveryBinder.java

```diff
@@ -84,11 +96,12 @@ public <T extends ServiceAnnouncement> void bindServiceAnnouncement(Class<? exte
         serviceAnnouncementBinder.addBinding().toProvider(announcementProviderClass);
     }
 
-    public ServiceAnnouncementBuilder bindHttpAnnouncement(String type)
+    public HttpServiceAnnouncementBuilder bindHttpAnnouncement(String type)
     {
-        ServiceAnnouncementBuilder serviceAnnouncementBuilder = serviceAnnouncement(type);
-        bindServiceAnnouncement(new HttpAnnouncementProvider(serviceAnnouncementBuilder));
-        return serviceAnnouncementBuilder;
+        Named annotation = named("http-announcement.%s.%s".formatted(type, randomUUID()));
```

Comment:
Do we need the generated annotation?

Reply:
how do you suggest to avoid it?

Reply:
I don't understand this code at all, so I don't know what you are trying to achieve here :)

Reply:
I think this should be idempotent rather than random, which would allow multiple modules to independently add bindings for the same announcement. Rather than using `Named`, we should have a `HttpAnnouncementType` annotation and implementation, like we have with `ServiceType` and `ServiceTypeImpl`.

Reply:
Slight modification of `io.airlift.discovery.client.TestHttpAnnouncementBinder#testHttpAnnouncementWithCustomProperties`:

```java
@Test
public void testHttpAnnouncementWithCustomProperties()
{
    final StaticAnnouncementHttpServerInfoImpl httpServerInfo = new StaticAnnouncementHttpServerInfoImpl(
            URI.create("http://127.0.0.1:4444"),
            URI.create("http://example.com:4444"),
            URI.create("https://127.0.0.1:4444"),
            URI.create("https://example.com:4444"));

    Injector injector = Guice.createInjector(
            new TestingDiscoveryModule(),
            binder -> {
                binder.bind(AnnouncementHttpServerInfo.class).toInstance(httpServerInfo);
                discoveryBinder(binder).bindHttpAnnouncement("apple").addProperty("a", "apple");
            },
            // Bind again to add new property
            binder -> {
                binder.bind(AnnouncementHttpServerInfo.class).toInstance(httpServerInfo);
                discoveryBinder(binder).bindHttpAnnouncement("apple").addProperty("b", "banana");
            });

    ServiceAnnouncement announcement = serviceAnnouncement("apple")
            .addProperty("a", "apple")
            .addProperty("b", "banana")
            .addProperty("http", httpServerInfo.getHttpUri().toASCIIString())
            .addProperty("http-external", httpServerInfo.getHttpExternalUri().toASCIIString())
            .addProperty("https", httpServerInfo.getHttpsUri().toASCIIString())
            .addProperty("https-external", httpServerInfo.getHttpsExternalUri().toASCIIString())
            .build();

    Set<ServiceAnnouncement> announcements = injector.getInstance(Key.get(new TypeLiteral<Set<ServiceAnnouncement>>() {
    }));

    assertAnnouncement(announcements, announcement);
}
```

This doesn't pass in current master, as it creates two announcements, not one.
The behavior is retained by this PR.

Reply:
Hmm, I see how this works now. In the existing code, `bindHttpAnnouncement()` creates `ServiceAnnouncementBuilder` which has a UUID, so we get a unique announcement for each call.